File: Python/src/tls.py
# Author: Kaden Daya <kaden.daya@icloud.com>
# Part of the betanet project by Raden Dev Team

import socket
import ssl

class client:
    def __init__(self, host, port):
        self.host = host
        self.port = port
        self.error = None
        self.socket = None
        
        try:
            self.context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
            self.context.check_hostname = False
            self.context.verify_mode = ssl.CERT_NONE
            self.context.minimum_version = ssl.TLSVersion.TLSv1_3
            self.context.maximum_version = ssl.TLSVersion.TLSv1_3
            
            self.context.set_alpn_protocols(['h2', 'http/1.1'])
            
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket = self.context.wrap_socket(self.socket, server_hostname=self.host)
        except ssl.SSLError as e:
            self.error = f"SSL context creation error: {e}"
            return
        except Exception as e:
            self.error = f"Client initialization error: {e}"
            return

    def connect(self):
        if self.error:
            return False, self.error
            
        try:
            self.socket.connect((self.host, self.port))
            return True, f"Connected to {self.host}:{self.port}"
        except ConnectionRefusedError:
            return False, f"Connection refused to {self.host}:{self.port} - server may not be running"
        except ssl.SSLError as e:
            return False, f"SSL/TLS error during connection: {e}"
        except socket.error as e:
            return False, f"Socket error during connection: {e}"
        except Exception as e:
            return False, f"Unexpected error during connection: {e}"

    def send(self, data):
        if self.error:
            return False, self.error
            
        try:
            if isinstance(data, str):
                return False, "Data must be bytes"
            self.socket.sendall(len(data).to_bytes(4, 'big') + data)
            return True, "Data sent successfully"
        except BrokenPipeError:
            return False, "Connection broken - cannot send data"
        except ConnectionResetError:
            return False, "Connection reset - cannot send data"
        except ssl.SSLError as e:
            return False, f"SSL/TLS error during send: {e}"
        except socket.error as e:
            return False, f"Socket error during send: {e}"
        except Exception as e:
            return False, f"Unexpected error during send: {e}"

    def receive(self):
        if self.error:
            return False, self.error
        try:
            hdr = bytearray()
            while len(hdr) < 4:
                chunk = self.socket.recv(4 - len(hdr))
                if not chunk:
                    return False, "Connection closed while reading length"
                hdr.extend(chunk)
            length = int.from_bytes(hdr, 'big')

            buf = bytearray()
            while len(buf) < length:
                chunk = self.socket.recv(length - len(buf))
                if not chunk:
                    return False, "Connection closed while reading payload"
                buf.extend(chunk)

            return True, bytes(buf)
        except ConnectionResetError:
            return False, "Connection reset by server"
        except ssl.SSLError as e:
            return False, f"SSL/TLS error during receive: {e}"
        except socket.error as e:
            return False, f"Socket error during receive: {e}"
        except Exception as e:
            return False, f"Unexpected error during receive: {e}"

    def close(self):
        if self.socket:
            try:
                self.socket.close()
                return True, "Client connection closed"
            except Exception as e:
                return False, f"Error closing client: {e}"
        return True, "No connection to close"

class server:
    def __init__(self, host, port, certfile, keyfile):
        self.host = host
        self.port = port
        self.error = None
        self.socket = None
        self.conn = None
        self.addr = None
        
        try:
            self.context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
            self.context.load_cert_chain(certfile=certfile, keyfile=keyfile)
            self.context.minimum_version = ssl.TLSVersion.TLSv1_3
            self.context.maximum_version = ssl.TLSVersion.TLSv1_3
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket = self.context.wrap_socket(self.socket, server_side=True)
            self.socket.bind((self.host, self.port))
        except FileNotFoundError as e:
            self.error = f"Certificate or key file not found: {e}"
            return
        except ssl.SSLError as e:
            self.error = f"SSL context creation error: {e}"
            return
        except PermissionError:
            self.error = f"Permission denied binding to {self.host}:{self.port}"
            return
        except socket.error as e:
            self.error = f"Socket error during server initialization: {e}"
            return
        except Exception as e:
            self.error = f"Server initialization error: {e}"
            return

    def start(self):
        if self.error:
            return False, self.error
            
        try:
            self.socket.listen(1)
            return True, f"Server is listening on {self.host}:{self.port}"
        except socket.error as e:
            return False, f"Socket error starting server: {e}"
        except ssl.SSLError as e:
            return False, f"SSL/TLS error starting server: {e}"
        except Exception as e:
            return False, f"Unexpected error starting server: {e}"

    def stop(self):
        if self.socket:
            try:
                self.socket.close()
                return True, "Server stopped"
            except Exception as e:
                return False, f"Error stopping server: {e}"
        return True, "No server to stop"

    def accept(self):
        if self.error:
            return False, self.error
            
        try:
            self.conn, self.addr = self.socket.accept()
            return True, f"Connected by {self.addr}", [self.conn, self.addr]
        except socket.error as e:
            return False, f"Socket error accepting connection: {e}"
        except ssl.SSLError as e:
            return False, f"SSL/TLS error accepting connection: {e}"
        except Exception as e:
            return False, f"Unexpected error accepting connection: {e}"

    def send(self, data):
        if self.error or not self.conn:
            return False, "No active connection"
        try:
            self.conn.sendall(len(data).to_bytes(4, 'big') + data)
            return True, "Data sent successfully"
        except BrokenPipeError:
            return False, "Connection broken - cannot send data"
        except ConnectionResetError:
            return False, "Connection reset - cannot send data"
        except ssl.SSLError as e:
            return False, f"SSL/TLS error during send: {e}"
        except socket.error as e:
            return False, f"Socket error during send: {e}"
        except Exception as e:
            return False, f"Unexpected error during send: {e}"

    def receive(self):
        if self.error or not self.conn:
            return False, "No active connection"
            
        try:
            hdr = bytearray()
            while len(hdr) < 4:
                chunk = self.conn.recv(4 - len(hdr))
                if not chunk:
                    return False, "Connection closed while reading length"
                hdr.extend(chunk)
            length = int.from_bytes(hdr, 'big')

            # Read payload
            buf = bytearray()
            while len(buf) < length:
                chunk = self.conn.recv(length - len(buf))
                if not chunk:
                    return False, "Connection closed while reading payload"
                buf.extend(chunk)

            return True, bytes(buf)
        except ConnectionResetError:
            return False, "Connection reset by client"
        except ssl.SSLError as e:
            return False, f"SSL/TLS error during receive: {e}"
        except socket.error as e:
            return False, f"Socket error during receive: {e}"
        except Exception as e:
            return False, f"Unexpected error during receive: {e}"

    def close(self):
        if self.conn:
            try:
                self.conn.close()
                self.conn = None
                self.addr = None
                return True, "Client connection closed"
            except Exception as e:
                return False, f"Error closing client connection: {e}"
        return True, "No client connection to close"

File: Rust/src/tls.rs
// Author: Kaden Daya <kaden.daya@icloud.com>
// Part of the betanet project by Raden Dev Team

use std::fs::File;
use std::io::{BufReader, Read, Write};
use std::net::{TcpStream, TcpListener};
use std::sync::Arc;

use rustls::{ServerConfig, ServerConnection, Stream, ClientConfig, ClientConnection, RootCertStore};
use rustls::pki_types::{PrivateKeyDer, CertificateDer, ServerName};
use rustls_pemfile::{certs, pkcs8_private_keys};

pub struct Client {
    host: String,
    port: u16,
    socket: Option<TcpStream>,
    tls_conn: Option<ClientConnection>,
    last_response: Option<String>,
}

pub struct Server {
    host: String,
    port: u16,
    cert_file: String,
    key_file: String,
    listener: Option<TcpListener>,
}

impl Client {
    pub fn new(host: String, port: u16) -> Self {
        Client {
            host, port, socket: None, tls_conn: None, last_response: None
        }
    }
    
    pub fn connect(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let addr = format!("{}:{}", self.host, self.port);
        let stream = TcpStream::connect(&addr)?;
        stream.set_nodelay(true)?;
        self.socket = Some(stream);
        Ok(())
    }

    pub fn setup_tls(&mut self, cert_path: &str) -> Result<(), Box<dyn std::error::Error>> {
        // Load root certificates
        let f = File::open(cert_path)?;
        let mut reader = BufReader::new(f);
        let parsed: Vec<CertificateDer<'static>> = certs(&mut reader).collect::<Result<_, _>>()?;

        let mut roots = RootCertStore::empty();
        for cert in parsed {
            roots.add(cert)?;
        }

        let config = Arc::new(
            ClientConfig::builder()
                .with_root_certificates(roots)
                .with_no_client_auth(),
        );

        let server_name = ServerName::try_from("localhost")?;
        let tls_conn = ClientConnection::new(config, server_name)?;
        self.tls_conn = Some(tls_conn);

        Ok(())
    }

    pub fn send(&mut self, data: &[u8]) -> Result<(), Box<dyn std::error::Error>> {
        if let (Some(ref mut tls_conn), Some(ref mut socket)) = (&mut self.tls_conn, &mut self.socket) {
            let mut tls_stream = Stream::new(tls_conn, socket);
            tls_stream.write_all(data)?;
            tls_stream.flush()?;
            Ok(())
        } else if let Some(ref mut socket) = self.socket {
            socket.write_all(data)?;
            socket.flush()?;
            Ok(())
        } else {
            Err("No connection established".into())
        }
    }

    pub fn receive(&mut self, buffer: &mut [u8]) -> Result<usize, Box<dyn std::error::Error>> {
        if let (Some(ref mut tls_conn), Some(ref mut socket)) = (&mut self.tls_conn, &mut self.socket) {
            let mut tls_stream = Stream::new(tls_conn, socket);
            let bytes_read = tls_stream.read(buffer)?;
            if bytes_read > 0 {
                let response = String::from_utf8_lossy(&buffer[..bytes_read]);
                self.last_response = Some(response.to_string());
            }
            Ok(bytes_read)
        } else if let Some(ref mut socket) = self.socket {
            let bytes_read = socket.read(buffer)?;
            if bytes_read > 0 {
                let response = String::from_utf8_lossy(&buffer[..bytes_read]);
                self.last_response = Some(response.to_string());
            }
            Ok(bytes_read)
        } else {
            Err("No connection established".into())
        }
    }

    pub fn get_last_response(&self) -> Option<&String> {
        self.last_response.as_ref()
    }

    pub fn close(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        if let Some(tls_conn) = self.tls_conn.take() {
            drop(tls_conn);
        }
        if let Some(socket) = self.socket.take() {
            drop(socket);
        }
        Ok(())
    }
}

impl Server {
    pub fn new(host: String, port: u16, cert_file: String, key_file: String) -> Self {
        Server {
            host, port, cert_file, key_file, listener: None
        }
    }

    pub fn load_certificates(&self) -> Result<ServerConfig, Box<dyn std::error::Error>> {
        let cert_file = File::open(&self.cert_file)?;
        let mut cert_reader = BufReader::new(cert_file);
        let cert_chain: Vec<_> = certs(&mut cert_reader).collect::<Result<_, _>>()?;
        
        let key_file = File::open(&self.key_file)?;
        let mut key_reader = BufReader::new(key_file);
        let mut keys: Vec<_> = pkcs8_private_keys(&mut key_reader).collect::<Result<_, _>>()?;
        
        if keys.is_empty() {
            return Err("No private keys found".into());
        }
        
        let config = ServerConfig::builder()
            .with_no_client_auth()
            .with_single_cert(cert_chain, PrivateKeyDer::Pkcs8(keys.remove(0)))?;
        
        Ok(config)
    }

    pub fn handle_client(mut stream: TcpStream, config: Arc<ServerConfig>) -> Result<String, Box<dyn std::error::Error>> {
        let mut tls_conn = ServerConnection::new(config)?;
    
        let mut tls_stream = Stream::new(&mut tls_conn, &mut stream);
        
        // Drive the TLS handshake to completion
        tls_stream.flush()?;
        
        let mut buffer = [0; 1024];
        let bytes_read = tls_stream.read(&mut buffer)?;
        
        if bytes_read > 0 {
            let request = String::from_utf8_lossy(&buffer[..bytes_read]);
            Ok(request.to_string())
        } else {
            Ok("No data received".to_string())
        }
    }

    pub fn send_response(&self, stream: &mut TcpStream, config: Arc<ServerConfig>, response: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut tls_conn = ServerConnection::new(config)?;
        let mut tls_stream = Stream::new(&mut tls_conn, stream);
        
        let response_bytes = response.as_bytes();
        tls_stream.write_all(response_bytes)?;
        tls_stream.flush()?;
        
        Ok(())
    }

    pub fn start(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let config = Arc::new(self.load_certificates()?);
        let listener = TcpListener::bind(format!("{}:{}", self.host, self.port))?;
        self.listener = Some(listener);
        
        if let Some(ref listener) = self.listener {
            for stream in listener.incoming() {
                match stream {
                    Ok(stream) => {
                        let config = config.clone();
                        
                        std::thread::spawn(move || {
                            match Self::handle_client(stream, config) {
                                Ok(request) => {
                                },
                                Err(e) => eprintln!("TLS connection error: {}", e),
                            }
                        });
                    }
                    Err(e) => {
                        eprintln!("Connection error: {}", e);
                    }
                }
            }
        }
        
        Ok(())
    }

    pub fn stop(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        if let Some(listener) = self.listener.take() {
            drop(listener);
        }
        Ok(())
    }

    pub fn accept(&mut self) -> Result<TcpStream, Box<dyn std::error::Error>> {
        if let Some(ref listener) = self.listener {
            match listener.accept() {
                Ok((stream, _addr)) => Ok(stream),
                Err(e) => Err(e.into())
            }
        } else {
            Err("Server not started".into())
        }
    }
}